src/util/ALE.sol

```diff
@@ -0,0 +1,459 @@
+        address indexed market,
+        address indexed account,
+        uint256 dolaFlashMinted,
+        uint256 collateralBought
+    );
+
+    event LeverageDown(
+        address indexed market,
+        address indexed account,
+        uint256 dolaFlashMinted,
+        uint256 collateralSold
+    );
+
+    event NewMarket(address indexed market, address indexed buySellToken, address collateral, address indexed helper);
+
+    event NewHelper(address indexed market, address indexed helper);
+
+    // Mapping of market to Market structs
+    // NOTE: in normal cases sellToken/buyToken is the collateral token,
+    // in other cases it could be different (eg. st-yCRV is collateral, yCRV is the token to be swapped from/to DOLA)
+    mapping(address => Market) public markets;
+
+    modifier dolaSupplyUnchanged() {
+        uint256 totalSupply = dola.totalSupply();
+        _;
+        if (totalSupply != dola.totalSupply())
+            revert TotalSupplyChanged(totalSupply, dola.totalSupply());
+    }
+
+    constructor(
+        address _exchangeProxy,
+        address _pool
+    ) Ownable(msg.sender) CurveDBRHelper(_pool) {
+        exchangeProxy = payable(address(_exchangeProxy));
+    }
+
+    function setExchangeProxy(address _exchangeProxy) external onlyOwner {
+        if (_exchangeProxy == address(0)) revert InvalidProxyAddress();
+        exchangeProxy = payable(_exchangeProxy);
+    }
+
+    /// @notice Set the market for a collateral token
+    /// @param _buySellToken The token which will be bought/sold (usually the collateral token), probably underlying if there's a helper
+    /// @param _market The market contract
+    /// @param _collateral The collateral token
+    /// @param _helper Optional helper contract to transform collateral to buySelltoken and viceversa
+    function setMarket(
+        address _market,
+        address _buySellToken,
+        address _collateral,
+        address _helper
+    ) external onlyOwner {
+        markets[_market].buySellToken = IERC20(_buySellToken);
+        markets[_market].collateral = IERC20(_collateral);
+        IERC20(_buySellToken).approve(_market, type(uint256).max);
+
+        if (_buySellToken != _collateral) {
+            IERC20(_collateral).approve(_market, type(uint256).max);
+        }
+
+        if (_helper != address(0)) {
+            markets[_market].helper = ITransformHelper(_helper);
+            IERC20(_buySellToken).approve(_helper, type(uint256).max);
+            IERC20(_collateral).approve(_helper, type(uint256).max);
+        }
+
+        emit NewMarket(_market, _buySellToken, _collateral, _helper);
+    }
+
+    /// @notice Update the helper contract
+    /// @param _market The market we want to update the helper contract for
+    /// @param _helper The helper contract
+    function updateMarketHelper(
+        address _market,
+        address _helper
+    ) external onlyOwner {
+        address oldHelper = address(markets[_market].helper);
```

Comment:
Suggestion:
Add validation to revert if market is not already setup. This enforces that the user supplied `market` is not an invalid or accidental value.